### PR TITLE
Tango YAML - translate XMI, fandago or running Tango device to YAML

### PR DESCRIPTION
**Pull request by SKAJohanVenter**
To aid in verification of device interfaces and documentation (like ICDs) a simplified format that is both human and machine readable is required.

This PR adds a `tango-yaml` tool that makes the generation of this file easier. Details of its use and examples of the format is available below in `doc/intro.rst`

**TODO** - Add more details about commands and attributes.

- [x] Code meets our [python style guidelines](https://docs.google.com/document/d/1aZoIyR9tz5rCWr2qJKuMTmKp2IzHlFjrCFrpDDHFypM/edit?usp=sharing)?
- [x] Unit tested (coded, passed, included)?
- [x] Requested at least 2 reviewers?
- [x] Commented code, particularly in hard-to-understand areas?
- [x] Made corresponding changes to the documentation (e.g. Python documentation, System Engineering Documentation, version description updates, README file, etc)?

JIRA: [SAR-112](https://jira.skatelescope.org/browse/SAR-112)

### CODE REVIEW

**Comment by pswart** [comment]
Lorenzo Pivetta asked (in SKA1 Pre-construction Phase) that the device minimum polling rate be documented in ICDs. This is maybe a good thing to include because it describes part of behaviour of the device.
_"The property min_poll_period (mS) defined a minimun polling period for the device."_ - see https://tango-controls.readthedocs.io/en/9.2.5/manual/07-advanced-features.html#device-polling

File: tango_simlib/tango_yaml_tools/base.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# BSD license - see LICENSE.txt for details                                             #
+#########################################################################################
+"""Module that contains the TangoToYAML class that parses a Tango device specification
+   file (xmi, fgo) or a running Tango device into YAML"""
+from pathlib import Path
+
+import yaml
+
+
+class TangoToYAML:
```
Comment (ajoubertza) [inline tango_simlib/tango_yaml_tools/base.py:13]: To still support Python 2.7:
```suggestion
class TangoToYAML(object):
```

File: tango_simlib/tango_yaml_tools/base.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#########################################################################################
+# Copyright 2020 SKA South Africa (http://ska.ac.za/)                                   #
+#                                                                                       #
+# BSD license - see LICENSE.txt for details                                             #
+#########################################################################################
+"""Module that contains the TangoToYAML class that parses a Tango device specification
+   file (xmi, fgo) or a running Tango device into YAML"""
+from pathlib import Path
```
Comment (ajoubertza) [inline tango_simlib/tango_yaml_tools/base.py:8]: Is this available in Python 2.7?

File: tango_simlib/tango_yaml_tools/base.py (hunk 1 above)
Comment (SKAJohanVenter) [reply tango_simlib/tango_yaml_tools/base.py:13]: @ajoubertza The CI pipeline installs and executes tests in both Python2 and Python3 environments.

Looks fine in 2.7.

```
Python 2.7.17 (default, Apr 15 2020, 17:20:14) 
[GCC 7.5.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> class O:
...     pass
... 
>>> o = O()
>>> o
<__main__.O instance at 0x7fd44122bf00>
```

File: tango_simlib/tango_yaml_tools/base.py (hunk 2 above)
Comment (SKAJohanVenter) [reply tango_simlib/tango_yaml_tools/base.py:8]: Yes, it's added in the `setup.py`, installed in Python2 and ignored in Python3

File: tango_simlib/tango_yaml_tools/base.py (hunk 1 above)
Comment (samueltwum1) [reply tango_simlib/tango_yaml_tools/base.py:13]: Interesting. Didnt know that works in py2.7

File: tango_simlib/tango_yaml_tools/main.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    """
+    if "xmi_file" in args:
+        return TangoToYAML(XmiParser).build_yaml_from_file(args.xmi_file.name)
+    if "fandango_file" in args:
+        return TangoToYAML(FP).build_yaml_from_file(args.fandango_file.name)
+    if "tango_device_name" in args:
+        return TangoToYAML(TangoDeviceParser).build_yaml_from_device(
+            args.tango_device_name
+        )
+    return ""
```
Comment (samueltwum1) [inline tango_simlib/tango_yaml_tools/main.py:39]: I didnt know that an if on empty string evaluates to false

File: tango_simlib/utilities/tango_device_parser.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                "disp_level": command.disp_level.name,
+            }
+
+        for prop in self.device_proxy.get_property_list("*"):
+            self.data_dict["meta"]["properties"][prop] = {"name": prop}
+
+    def get_device_attribute_metadata(self):
+        """Returns the attributes.
+
+
```
Comment (kmadisa) [inline tango_simlib/utilities/tango_device_parser.py:80]: Just some minor thing, the extra whitespace.

File: tango_simlib/tests/test_yaml_builder.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+def validate_basic_structure(yaml_dict):
+    """Do a basic check of the structure
+    """
+    assert len(yaml_dict) == 1
+    assert "meta" in yaml_dict[0]
+    assert "class" in yaml_dict[0]
+    assert "properties" in yaml_dict[0]["meta"]
+    assert "attributes" in yaml_dict[0]["meta"]
+
```
Comment (kmadisa) [inline tango_simlib/tests/test_yaml_builder.py:30]: Aren't going to check for the commands key here?

Comment (SKAJohanVenter) [reply tango_simlib/tests/test_yaml_builder.py:30]: I should 😄

File: tango_simlib/tango_yaml_tools/base.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        for attr in self.parser.get_device_attribute_metadata().values():
+            attr_data = {"name": attr["name"]}
+            if "data_type" in attr:
+                attr_data["data_type"] = attr["data_type"].name
+            if "data_format" in attr:
+                attr_data["data_format"] = attr["data_format"].name
+            if "disp_level" in attr:
+                attr_data["disp_level"] = attr["disp_level"].name
+            # pylint insists on this spacing
+            for key in [
```
Comment (samueltwum1) [inline tango_simlib/tango_yaml_tools/base.py:62]: Probably much more descriptive to assign the list as `attr_props` to make it more descriptive

Comment (SKAJohanVenter) [reply tango_simlib/tango_yaml_tools/base.py:62]: Attributes and Properties are different things in devices, so `attr_props` would be confusing

Comment (samueltwum1) [reply tango_simlib/tango_yaml_tools/base.py:62]: Yeah so aside the device's properties, the commands and attributes also have properties (e.g. the identifiers in the list). Takes getting used to but "Tango speaking" people will understand usage of `attr_props`. Still looks neat this way and probably better to leave it as is so any "non-tango" pair of eyes looking at this code dont get confused.

**Pull request merged**